File: packages/agent-runtime-py/ainative_agent_runtime/model_selector.py
"""
ainative-agent-runtime — ModelSelector (Python)
Built by AINative Dev Team
Refs #248

Selects the best LLM provider based on availability and task complexity.
"""

from __future__ import annotations

import time
from typing import Any, Dict, List, Optional


class ModelSelector:
    """
    Selects an LLM provider for a given task.

    Selection order:
        1. If complexity_threshold is set and task.metadata.complexity > threshold,
           prefer the last registered (cloud) provider.
        2. Otherwise prefer the first healthy provider (local).
        3. If no healthy provider, raise RuntimeError.

    Args:
        providers: List of LLM provider objects, each must have a `.name` attribute.
        complexity_threshold: Optional int/float. Tasks above this value route to cloud.
    """

    def __init__(
        self,
        providers: List[Any],
        complexity_threshold: Optional[float] = None,
    ) -> None:
        self._providers = providers
        self._complexity_threshold = complexity_threshold
        # Initially all providers assumed healthy
        self._health: Dict[str, bool] = {p.name: True for p in providers}

    # ─── select() ─────────────────────────────────────────────────────────────

    async def select(self, task: Dict[str, Any]) -> Any:
        if not self._providers:
            raise RuntimeError("No providers registered in ModelSelector")

        complexity = float(task.get("metadata", {}).get("complexity", 0) or 0)
        is_high_complexity = (
            self._complexity_threshold is not None
            and complexity > self._complexity_threshold
        )

        if is_high_complexity:
            cloud = self._providers[-1]
            if self._health.get(cloud.name, True):
                return cloud

        # Return first healthy provider
        for provider in self._providers:
            if self._health.get(provider.name, True):
                return provider

        raise RuntimeError("No healthy providers available")

    # ─── health_check() ───────────────────────────────────────────────────────

    async def health_check(self) -> List[Dict[str, Any]]:
        """Ping all providers and record their health status."""
        results = []

        for provider in self._providers:
            start = time.monotonic()
            healthy = True
            error: Optional[str] = None

            try:
                await provider.chat([{"role": "user", "content": "ping"}])
            except Exception as exc:
                healthy = False
                error = str(exc)

            latency_ms = (time.monotonic() - start) * 1000
            self._health[provider.name] = healthy

            results.append({
                "name": provider.name,
                "healthy": healthy,
                "latency_ms": latency_ms,
                "error": error,
            })

        return results

File: packages/agent-runtime/src/model-selector.ts
/**
 * @ainative/agent-runtime — ModelSelector
 * Built by AINative Dev Team
 * Refs #248
 *
 * Selects the best LLM provider based on:
 *   1. Local availability (health check result)
 *   2. Task complexity (metadata.complexity vs threshold)
 *   3. Fallback to next available provider
 */

import type { LLMProvider, AgentTask, ProviderHealth } from './types';

// ─── Config ───────────────────────────────────────────────────────────────────

export interface ModelSelectorConfig {
  providers: Array<LLMProvider & { name: string }>;
  /**
   * Tasks with metadata.complexity above this value are routed to the
   * last registered provider (assumed to be cloud). Default: none.
   */
  complexityThreshold?: number;
}

// ─── ModelSelector ───────────────────────────────────────────────────────────

export class ModelSelector {
  private readonly providers: Array<LLMProvider & { name: string }>;
  private readonly complexityThreshold: number | undefined;
  private healthMap: Map<string, boolean> = new Map();

  constructor(config: ModelSelectorConfig) {
    this.providers = config.providers;
    this.complexityThreshold = config.complexityThreshold;
    // Initially mark all providers as healthy (unknown)
    for (const p of this.providers) {
      this.healthMap.set(p.name, true);
    }
  }

  // ─── select() ─────────────────────────────────────────────────────────────

  /**
   * Choose the appropriate provider for a given task.
   *
   * Selection order:
   *   1. If complexityThreshold is set and task complexity exceeds it,
   *      prefer the last provider (cloud fallback).
   *   2. Otherwise prefer the first healthy provider (local).
   *   3. If no healthy provider found, throw.
   */
  async select(task: AgentTask): Promise<LLMProvider & { name: string }> {
    if (this.providers.length === 0) {
      throw new Error('No providers registered in ModelSelector');
    }

    const complexity = typeof task.metadata?.complexity === 'number'
      ? (task.metadata.complexity as number)
      : 0;

    const isHighComplexity =
      this.complexityThreshold !== undefined && complexity > this.complexityThreshold;

    if (isHighComplexity) {
      // Prefer the last registered (cloud) provider if it's healthy
      const cloud = this.providers[this.providers.length - 1];
      if (this.healthMap.get(cloud.name) !== false) return cloud;
    }

    // Find first healthy provider
    for (const provider of this.providers) {
      if (this.healthMap.get(provider.name) !== false) {
        return provider;
      }
    }

    throw new Error('No healthy providers available');
  }

  // ─── healthCheck() ────────────────────────────────────────────────────────

  /**
   * Ping all providers with a minimal request and record their health status.
   */
  async healthCheck(): Promise<ProviderHealth[]> {
    const results: ProviderHealth[] = [];

    for (const provider of this.providers) {
      const start = Date.now();
      let healthy = true;
      let error: string | undefined;

      try {
        await provider.chat([{ role: 'user', content: 'ping' }]);
      } catch (err) {
        healthy = false;
        error = err instanceof Error ? err.message : String(err);
      }

      const latencyMs = Date.now() - start;

      this.healthMap.set(provider.name, healthy);

      results.push({
        name: provider.name,
        healthy,
        latencyMs,
        error,
      });
    }

    return results;
  }
}
